frontend/gateway/grpcclient/client.go

```diff
@@ -395,8 +411,27 @@ func (c *grpcClient) BuildOpts() client.BuildOpts {
 }
 
 type reference struct {
-	id string
-	c  *grpcClient
+	llb.Output
+	c   *grpcClient
+	id  string
+	def *opspb.Definition
+}
+
+func newReference(c *grpcClient, ref *pb.Ref) (*reference, error) {
+	if len(ref.Ids) > 1 {
+		return nil, errors.Errorf("cannot create multi-result array reference")
+	}
+
+	if len(ref.Ids) != len(ref.Defs) {
+		return nil, errors.Errorf("reference ids and definitions mismatch length")
+	}
+
+	dop, err := llb.NewDefinitionOp(ref.Defs[0])
```

Comment:
Do this `[0]` selectors need to be protected to avoid panic (at least on old/misbehaving daemons)?

Reply:
I'll protect them. :+1: